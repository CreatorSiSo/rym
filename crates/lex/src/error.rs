use crate::Lexer;
use std::fmt::Display;
use std::num::{ParseFloatError, ParseIntError};

#[derive(Debug, PartialEq, Eq)]
pub enum LexerError {
	InvalidChar {
		msg: String,
		line: usize,
		col: usize,
	},
	ParseInt {
		msg: String,
		line: usize,
		col: usize,
	},
	ParseFloat {
		msg: String,
		line: usize,
		col: usize,
	},
}

impl LexerError {
	pub fn invalid_char<T>(lexer: &Lexer) -> Result<T, Self> {
		Err(Self::InvalidChar {
			msg: format!("Found invalid character `{}`", lexer.c),
			line: lexer.line,
			col: lexer.col,
		})
	}

	pub fn parse_int<T>(lexer: &Lexer, err: ParseIntError) -> Result<T, Self> {
		Err(Self::ParseInt {
			msg: format!("{} -> {err:?}", &lexer.source[lexer.start..lexer.current]),
			line: lexer.line,
			col: lexer.col,
		})
	}

	pub fn parse_float<T>(lexer: &Lexer, err: ParseFloatError) -> Result<T, Self> {
		Err(Self::ParseFloat {
			msg: format!("{} -> {err:?}", &lexer.source[lexer.start..lexer.current]),
			line: lexer.line,
			col: lexer.col,
		})
	}
}

impl Display for LexerError {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			LexerError::InvalidChar { msg, line, col } => {
				writeln!(f, "Error: {msg}")?;
				writeln!(f, "       {line}:{col}")
			}
			LexerError::ParseInt { msg, line, col } => {
				writeln!(f, "Error: Could not parse integer {msg}")?;
				writeln!(f, "       {line}:{col}")
			}
			LexerError::ParseFloat { msg, line, col } => {
				writeln!(f, "Error: Could not parse float {msg}")?;
				writeln!(f, "       {line}:{col}")
			}
		}
	}
}
