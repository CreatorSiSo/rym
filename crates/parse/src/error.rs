use std::fmt::Display;

use ast::{Spanned, Token};
use colored::Colorize;

#[derive(Debug, PartialEq)]
pub enum ParseError {
	TokenMismatch(Spanned<Token>, String),
	InvalidAssignmentTarget(Spanned<Token>),
}

impl ParseError {
	pub(super) fn token_mismatch<T>(token: &Spanned<Token>, msg: &str) -> Result<T, Self> {
		Err(ParseError::TokenMismatch(token.clone(), msg.into()))
	}
}

// TODO: Improve error messages by using Spanned properly
impl Display for ParseError {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		let err = "Error:".red().bold();

		match self {
			ParseError::TokenMismatch(Spanned(token, span), msg) => {
				writeln!(f, "{err}	{msg} got `{:?}`", token.typ)?;
				writeln!(f, "Span:	{:?}", span)
			}
			ParseError::InvalidAssignmentTarget(Spanned(token, span)) => {
				writeln!(f, "{err}	Invalid assignment target `{:?}`", token.typ)?;
				writeln!(f, "Span:	{:?}", span)
			}
		}
	}
}
