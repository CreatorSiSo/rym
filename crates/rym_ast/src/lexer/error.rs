use crate::Lexer;
use std::fmt::Display;
use std::num::{ParseFloatError, ParseIntError};

#[derive(Debug)]
pub enum LexerError {
	UnexpectedChar {
		msg: String,
		line: usize,
		col: usize,
	},
	ParseInt {
		msg: String,
		line: usize,
		col: usize,
	},
	ParseFloat {
		msg: String,
		line: usize,
		col: usize,
	},
}

impl LexerError {
	pub fn unexpected_char<T>(lexer: &Lexer) -> Result<T, Self> {
		Err(Self::UnexpectedChar {
			msg: format!("Unexpected character `{}`", lexer.c),
			line: lexer.line,
			col: lexer.col,
		})
	}

	pub fn parse_int<T>(lexer: &Lexer, err: ParseIntError) -> Result<T, Self> {
		Err(Self::ParseInt {
			msg: format!("{} -> {err:?}", &lexer.source[lexer.start..lexer.current]),
			line: lexer.line,
			col: lexer.col,
		})
	}

	pub fn parse_float<T>(lexer: &Lexer, err: ParseFloatError) -> Result<T, Self> {
		Err(Self::ParseFloat {
			msg: format!("{} -> {err:?}", &lexer.source[lexer.start..lexer.current]),
			line: lexer.line,
			col: lexer.col,
		})
	}
}

impl Display for LexerError {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			LexerError::UnexpectedChar { msg, line, col } => {
				writeln!(f, "Error: {msg}")?;
				writeln!(f, "       {line}:{col}")
			}
			LexerError::ParseInt { msg, line, col } => {
				writeln!(f, "Error: Could not parse integer {msg}")?;
				writeln!(f, "       {line}:{col}")
			}
			LexerError::ParseFloat { msg, line, col } => {
				writeln!(f, "Error: Could not parse float {msg}")?;
				writeln!(f, "       {line}:{col}")
			}
		}
	}
}
