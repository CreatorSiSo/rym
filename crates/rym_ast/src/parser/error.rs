use crate::token::*;
use std::fmt::Display;

pub enum ParserError<'src> {
	TokenMismatch { token: Token<'src>, msg: String },
	InvalidAssignmentTarget { token: Token<'src> },
}

impl<'src> ParserError<'src> {
	pub(super) fn token_mismatch<T>(token: Token<'src>, msg: &str) -> Result<T, Self> {
		Err(ParserError::TokenMismatch {
			token,
			msg: msg.into(),
		})
	}
}

impl Display for ParserError<'_> {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			ParserError::TokenMismatch { token, msg } => {
				writeln!(f, "Error: {msg} got `{:?}`", token.value)?;
				writeln!(f, "Position: {}", token.start)
			}
			ParserError::InvalidAssignmentTarget { token } => {
				writeln!(f, "Error: Invalid assignment target `{:?}`", token.value)?;
				writeln!(f, "Position: {}", token.start)
			}
		}
	}
}
