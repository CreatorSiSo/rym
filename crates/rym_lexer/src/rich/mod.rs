use rym_unescape::unquote;
use Mapped::{Multi, Single};

mod test;
mod token;
pub use token::*;

use crate::primitive;

type Span = std::ops::Range<usize>;

#[derive(Clone, Debug)]
pub struct Lexer<'a> {
	/// Absolute offset within the source of the current character.
	pos: usize,
	/// Source text to tokenize.
	src: &'a str,
	/// Cursor for getting lexer tokens.
	primitive_lexer: primitive::Lexer<'a>,
}

impl<'a> Lexer<'a> {
	pub fn new(src: &'a str) -> Self {
		Self { pos: 0, src, primitive_lexer: primitive::Lexer::new(src) }
	}

	fn bump(&mut self) -> Option<(primitive::TokenKind, Span)> {
		self.primitive_lexer.next_token().map(|token| {
			let start_pos = self.pos;
			self.pos += token.len as usize;
			(token.kind, start_pos..self.pos)
		})
	}

	fn peek(&self) -> Option<(primitive::TokenKind, Span)> {
		self.primitive_lexer.clone().next_token().map(|token| {
			let end_pos = self.pos + token.len as usize;
			(token.kind, self.pos..end_pos)
		})
	}

	fn src_from_span(&self, span: &Span) -> &'a str {
		&self.src[span.start..span.end]
	}

	fn map_next(
		&mut self,
		span: &mut Span,
		map_fn: impl Fn(Option<primitive::TokenKind>) -> Mapped,
	) -> Token {
		let mut end = span.end;
		match if let Some((kind, end_span)) = self.peek() {
			end = end_span.end;
			map_fn(Some(kind))
		} else {
			map_fn(None)
		} {
			Mapped::Multi(kind) => {
				*span = span.start..end;
				self.bump();
				kind
			}
			Mapped::Single(kind) => kind,
		}
	}
}

enum Mapped {
	Single(Token),
	Multi(Token),
}

impl Iterator for Lexer<'_> {
	type Item = (Token, Span);

	fn next(&mut self) -> Option<Self::Item> {
		while let Some((primitive_kind, mut span)) = self.bump() {
			let kind = match primitive_kind {
				primitive::TokenKind::Whitespace => {
					if self.src_from_span(&span).contains('\n') {
						Token::Newline
					} else {
						continue;
					}
				}
				primitive::TokenKind::LineComment => continue,
				primitive::TokenKind::BlockComment { terminated } => {
					if terminated {
						continue;
					}
					// TODO: Special reporting for nested block comments
					// self.handler.emit(
					// 	Diagnostic::new_spanned(Level::Error, "Unterminated block comment", span)
					// 		.sub_diagnostic(
					// 			Level::Note,
					// 			None,
					// 			"Missing trailing `*/` to terminate the block comment",
					// 		),
					// );
					continue;
				}

				// Punctuation
				primitive::TokenKind::Semi => Token::Semi,
				primitive::TokenKind::Colon => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Colon) => Multi(Token::ColonColon),
					_ => Single(Token::Colon),
				}),
				primitive::TokenKind::Dot => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Dot) => Multi(Token::DotDot),
					_ => Single(Token::Dot),
				}),
				primitive::TokenKind::Comma => Token::Comma,

				// Operator like
				primitive::TokenKind::Or => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Or) => Multi(Token::OrOr),
					_ => Single(Token::Or),
				}),
				primitive::TokenKind::And => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::And) => Multi(Token::AndAnd),
					_ => Single(Token::And),
				}),
				primitive::TokenKind::Plus => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::PlusEq),
					_ => Single(Token::Plus),
				}),
				primitive::TokenKind::Minus => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::MinusEq),
					Some(primitive::TokenKind::GreaterThan) => Multi(Token::ThinArrow),
					_ => Single(Token::Minus),
				}),
				primitive::TokenKind::Star => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::StarEq),
					_ => Single(Token::Star),
				}),
				primitive::TokenKind::Slash => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::SlashEq),
					_ => Single(Token::Slash),
				}),
				primitive::TokenKind::Percent => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::PercentEq),
					_ => Single(Token::Percent),
				}),
				primitive::TokenKind::Eq => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::EqEq),
					Some(primitive::TokenKind::GreaterThan) => Multi(Token::FatArrow),
					_ => Single(Token::Eq),
				}),
				primitive::TokenKind::Bang => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::BangEq),
					_ => Single(Token::Bang),
				}),
				primitive::TokenKind::LessThan => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::LessThanEq),
					_ => Single(Token::LessThan),
				}),
				primitive::TokenKind::GreaterThan => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::GreaterThanEq),
					_ => Single(Token::GreaterThan),
				}),

				// Delimiter
				primitive::TokenKind::OpenParen => Token::OpenParen,
				primitive::TokenKind::CloseParen => Token::CloseParen,
				primitive::TokenKind::OpenBrace => Token::OpenBrace,
				primitive::TokenKind::CloseBrace => Token::CloseBrace,
				primitive::TokenKind::OpenBracket => Token::OpenBracket,
				primitive::TokenKind::CloseBracket => Token::CloseBracket,

				// Indentifier or Keyword
				primitive::TokenKind::Ident => {
					let name = self.src_from_span(&span);
					match KEYWORDS_MAP.0.iter().position(|kw| *kw == name) {
						Some(pos) => KEYWORDS_MAP.1[pos].clone(),
						None => Token::Ident(name.to_string()),
					}
				}

				primitive::TokenKind::Integer => Token::Int(
					match self
						.src_from_span(&span)
						.chars()
						.filter(|c| c != &'_')
						.collect::<String>()
						.parse::<i64>()
					{
						Ok(int) => int,
						Err(err) => unreachable!(
							"Internal Error: Could not parse <i64> from {0}: `{err}`",
							self.src_from_span(&span)
						),
					},
				),
				primitive::TokenKind::Float => {
					let Some((l, r)) = self.src_from_span(&span).split_once('.') else {
							unreachable!("Internal Error: Float literal does not contain a '.'")
						};
					let l_val: u32 = l.parse().expect(&format!(
						"Internal Error: Left hand side of float literal has invalid value `{l}`"
					));
					let r_val: u32 = r.parse().unwrap_or(0);
					Token::Float(l_val, r_val)
				}
				primitive::TokenKind::Char { terminated } => {
					if !terminated {
						todo!();
						// TODO
						// self.handler.emit(
						// 	Diagnostic::new_spanned(Level::Error, "Unterminated character literal", span)
						// 		.sub_diagnostic(
						// 			Level::Note,
						// 			None,
						// 			"Missing trailing `'` to terminate the character literal",
						// 		),
						// );
						// continue;
					}
					let string = match unquote(self.src_from_span(&span)) {
						Ok(string) => string,
						Err(err) => todo!("Unquote error: {err}"),
					};
					Token::Char(match string.parse::<char>() {
						Ok(char) => char,
						Err(_err) => {
							todo!();
							// TODO
							// self.handler.emit(Diagnostic::new_spanned(
							// 	Level::Error,
							// 	format!("Could not parse <char>: {err}"),
							// 	span,
							// ));
							// continue;
						}
					})
				}
				primitive::TokenKind::String { terminated } => {
					if !terminated {
						// TODO
						// self.handler.emit(
						// 	Diagnostic::new_spanned(Level::Error, "Unterminated string literal", span)
						// 		.sub_diagnostic(
						// 			Level::Note,
						// 			None,
						// 			"Missing trailing `\"` to terminate the string literal",
						// 		),
						// );
						continue;
					}
					let string = match unquote(self.src_from_span(&span)) {
						Ok(string) => string,
						Err(_err) => {
							// TODO
							// self.handler.emit(Diagnostic::new_spanned(Level::Error, err.to_string(), span));
							continue;
						}
					};
					Token::String(string)
				}

				primitive::TokenKind::Unkown => {
					// TODO
					// self.handler.emit(Diagnostic::new_spanned(Level::Error, "Invalid character", span));
					continue;
				}
				primitive::TokenKind::At
				| primitive::TokenKind::Caret
				| primitive::TokenKind::Dollar
				| primitive::TokenKind::Pound
				| primitive::TokenKind::Tilde
				| primitive::TokenKind::Question => {
					// TODO
					// self.handler.emit(Diagnostic::new_spanned(Level::Error, "Reserved character", span));
					continue;
				}
			};

			return Some((kind, span));
		}
		None
	}
}
