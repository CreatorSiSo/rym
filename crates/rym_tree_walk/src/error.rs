use rym_ast::{Literal, UnaryOp};

#[derive(Debug)]
pub enum RuntimeError {
	ForbiddenType(String),
	TypeMismatch(String),
	UndeclaredVar(String),
	Assignment(String),
	DivideByZero,
}

// TODO: Print line number as well
impl RuntimeError {
	pub fn assignment<T>(name: &str, value: Literal) -> Result<T, Self> {
		Err(Self::Assignment(format!(
			"Cannot assign `{value}` to constant {name}"
		)))
	}

	pub fn undeclared_var<T>(name: &str) -> Result<T, Self> {
		Err(Self::UndeclaredVar(format!(
			"Variable `{name}` has not been declared"
		)))
	}

	pub fn unary<T>(op: &UnaryOp, right: Literal) -> Result<T, Self> {
		Err(Self::ForbiddenType(format!(
			"Cannot apply unary operator `{}` to `{}`",
			match op {
				UnaryOp::Neg => "-",
				UnaryOp::Not => "!",
			},
			right.to_type_string(),
		)))
	}

	pub fn expected<T>(type_str: &str, value: Literal) -> Result<T, Self> {
		Err(Self::TypeMismatch(format!(
			"Expected `{type_str}` got {}",
			value.to_type_string()
		)))
	}

	pub fn comparison<T>(left: Literal, right: Literal) -> Result<T, Self> {
		Err(Self::TypeMismatch(format!(
			"Cannot compare `{}` with `{}`",
			left.to_type_string(),
			right.to_type_string(),
		)))
	}

	pub fn addition<T>(left: Literal, right: Literal) -> Result<T, Self> {
		Err(Self::TypeMismatch(format!(
			"Cannot add `{}` to `{}`",
			left.to_type_string(),
			right.to_type_string(),
		)))
	}

	pub fn substraction<T>(left: Literal, right: Literal) -> Result<T, Self> {
		Err(Self::TypeMismatch(format!(
			"Cannot substract `{}` from `{}`",
			right.to_type_string(),
			left.to_type_string(),
		)))
	}

	pub fn multiplication<T>(left: Literal, right: Literal) -> Result<T, Self> {
		Err(Self::TypeMismatch(format!(
			"Cannot multiply `{}` by `{}`",
			left.to_type_string(),
			right.to_type_string(),
		)))
	}

	pub fn division<T>(left: Literal, right: Literal) -> Result<T, Self> {
		Err(Self::TypeMismatch(format!(
			"Cannot divide `{}` by `{}`",
			left.to_type_string(),
			right.to_type_string(),
		)))
	}
}
