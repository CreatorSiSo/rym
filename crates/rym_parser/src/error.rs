use crate::{ErrorAlias, Span, Token};
use chumsky::error::RichReason;
use std::fmt::Display;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum Label {
	Block,
	Break,
	Continue,
	Expression,
	Function,
	Group,
	Identifier,
	Literal,
	Loop,
	Module,
	Record,
	Return,
	Binding,
}

impl From<&Label> for &'static str {
	fn from(value: &Label) -> Self {
		match value {
			Label::Binding => "binding",
			Label::Block => "block",
			Label::Break => "break",
			Label::Continue => "continue",
			Label::Expression => "expression",
			Label::Function => "function",
			Label::Group => "group",
			Label::Identifier => "identifier",
			Label::Literal => "literal",
			Label::Loop => "loop",
			Label::Module => "module",
			Label::Record => "record",
			Label::Return => "return",
		}
	}
}

impl Display for Label {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		let data = <&'static str>::from(self);
		f.write_str(data)
	}
}

#[derive(Debug, PartialEq, Eq)]
pub struct ParseError<'a> {
	pub span: Span,
	pub reason: RichReason<'a, Token, Label>,
}

impl<'a> From<ErrorAlias<'a>> for ParseError<'a> {
	fn from(value: ErrorAlias<'a>) -> ParseError<'a> {
		fn deep_clone_sort_reason(reason: RichReason<Token, Label>) -> RichReason<Token, Label> {
			match reason {
				RichReason::ExpectedFound {
					mut expected,
					found,
				} => {
					expected.sort();
					RichReason::ExpectedFound { expected, found }
				}
				RichReason::Many(mut reasons) => {
					reasons.sort();
					RichReason::Many(reasons.into_iter().map(deep_clone_sort_reason).collect())
				}
				reason => reason,
			}
		}

		Self {
			span: value.span().clone(),
			reason: deep_clone_sort_reason(value.reason().clone()),
		}
	}
}
