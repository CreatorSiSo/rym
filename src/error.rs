use crate::{ErrorAlias, Span, Token};
use chumsky::error::RichReason;
use std::fmt::Display;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum Label {
	Block,
	Break,
	Continue,
	Expression,
	Function,
	Group,
	Identifier,
	Loop,
	Module,
	Record,
	Return,
	Binding,
}

impl From<&Label> for &'static str {
	fn from(value: &Label) -> Self {
		match value {
			Label::Binding => "binding",
			Label::Block => "block",
			Label::Break => "break",
			Label::Continue => "continue",
			Label::Expression => "expression",
			Label::Function => "function",
			Label::Group => "group",
			Label::Identifier => "identifier",
			Label::Loop => "loop",
			Label::Module => "module",
			Label::Record => "record",
			Label::Return => "return",
		}
	}
}

impl Display for Label {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		let data = <&'static str>::from(self);
		f.write_str(data)
	}
}

#[derive(Debug, PartialEq, Eq)]
pub struct ParseError<'a> {
	pub span: Span,
	pub reason: RichReason<'a, Token, Label>,
}

impl<'a> From<ErrorAlias<'a>> for ParseError<'a> {
	fn from(value: ErrorAlias<'a>) -> ParseError<'a> {
		fn deep_clone_sort_reason(reason: RichReason<Token, Label>) -> RichReason<Token, Label> {
			match reason {
				RichReason::ExpectedFound {
					mut expected,
					found,
				} => {
					expected.sort();
					RichReason::ExpectedFound { expected, found }
				}
				RichReason::Many(mut reasons) => {
					reasons.sort();
					RichReason::Many(reasons.into_iter().map(deep_clone_sort_reason).collect())
				}
				reason => reason,
			}
		}

		Self {
			span: value.span().clone(),
			reason: deep_clone_sort_reason(value.reason().clone()),
		}
	}
}

#[derive(Debug, PartialEq, Eq)]
pub struct ParseResult<'a, T>(pub Option<T>, pub Vec<ParseError<'a>>);

impl<'a, T> From<chumsky::ParseResult<T, ErrorAlias<'a>>> for ParseResult<'a, T> {
	fn from(value: chumsky::ParseResult<T, ErrorAlias<'a>>) -> Self {
		let (output, errors) = value.into_output_errors();
		let errors: Vec<_> = errors.into_iter().map(|err| err.into()).collect();
		Self(output, errors)
	}
}
