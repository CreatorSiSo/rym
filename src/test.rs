#![cfg(test)]

use super::*;
use chumsky::error::SimpleReason;
use chumsky::Stream;
use indoc::indoc;
use rym_lexer::rich::Lexer;

#[derive(Debug)]
#[allow(dead_code)]
struct Report<'a> {
	label: Option<&'a str>,
	span: Span,
	reason: &'a SimpleReason<Token, Span>,
	expected: Vec<&'a Option<Token>>,
	found: Option<&'a Token>,
}

impl<'a> From<&'a Simple<Token>> for Report<'a> {
	fn from(err: &'a Simple<Token>) -> Self {
		Report {
			label: err.label(),
			span: err.span(),
			reason: err.reason(),
			expected: {
				let mut expected = err.expected().collect::<Vec<_>>();
				expected.sort();
				expected
			},
			found: err.found(),
		}
	}
}

macro_rules! insta_assert_parser {
	($parser:expr; $($src:expr),+ $(,)?) => {
		$({
			let token_stream = Stream::from_iter(0..0, Lexer::new($src));
			let (ast, mut errors) = expr_parser().parse_recovery(token_stream);

			errors.sort_by(|l, r| l.span().start.cmp(&r.span().start));
			let reports: Vec<Report> = errors.iter().map(|err| Report::from(err)).collect();

			let snapshot = format!(
				"--- Input ---\n{}\n---\n\n{:#?}\n\n--- Errors ---\n{:#?}\n---",
				$src, &ast, &reports
			);
			insta::assert_snapshot!(&snapshot);
		})*
	};
}

#[test]
fn literal_expressions() {
	insta_assert_parser! {
		expr_parser();
		"823_472_340",
		"823_472_340_983_273_327",
		"1.0",
		"0.",
		"444.444",
		"9_999_999_999_999.999_999_999",
		"'a'",
		"'\\n'",
		"'\\t'",
		"'\\r'",
		r#" "abc" "#,
		r#" "Hello World!\n" "#,
	}
}

#[test]
fn record_expressions() {
	insta_assert_parser! {
		expr_parser();
		r#"Record { name: "Record", fields: todo() }"#,
		r#".{ name: "Record", fields: todo() }"#,
		r#".{ }"#,
	}
}

#[test]
fn simple_expressions() {
	insta_assert_parser! {
		expr_parser();
		"indentifier_1",
		"(grouped_ident)",
		"func_name(2, \"Do stuff!!\", true)",
		"(make_func())()()()",
		"(((wrapped)))()",
		"(if true then 1 else 2) + 3",
	}
}

#[test]
fn unary_expressions() {
	insta_assert_parser! {
		expr_parser();
		"-255",
		"-1.0",
		"-0.",
		"---3",

		"!true",
		"!false",
		"!!!true",
		"!(lhs == rhs)",
		"!lhs == rhs",
	}
}

#[test]
fn binary_expressions() {
	insta_assert_parser! {
		expr_parser();
		"lhs == rhs",
		"lhs != rhs",
		"lhs + rhs",
		"lhs - rhs",
		"lhs * rhs",
		"lhs / rhs",


		"1 + 2 - 3 * 4 / 5",
		"1 == 1 + 2 / 3 != 0",
	}
}

#[test]
fn block_expressions() {
	insta_assert_parser! {
		expr_parser();
		indoc!("
		{
			testing;
			testing;
		}"),
		"{ call_me(); and_me(); }",
		"{ }",
	}
}

#[test]
fn variables() {
	insta_assert_parser! {
		expr_parser();
		indoc!(r#"
		{
			const msg = "Hello World!\n";
			mut counter = 0;

			// TODO: count from 0 to 10 and print f"{counter}: Hello World!\n" every time
		}"#),
	}
}

#[test]
fn if_expressions() {
	insta_assert_parser! {
		expr_parser();
		r#"if true then { print("Hello Universe!"); } else { print("Hello World!"); }"#,
		r#"if (true == false) then ("Hello Universe!") else ("Hello World!")"#,
		r#"if true then "Hello Universe!" else "Hello World!""#,
		indoc!(r#"
			if true then "Hello Universe!"
			else "Hello World!""#),
		indoc!(r#"
			if stuff_goes_right(1, 2, 9) then {
				print("Hello Universe!");
				20;
			} else {
				print("Hello World!");
				10;
			}"#),
	}
}

#[test]
fn recover_group() {
	insta_assert_parser!(
		expr_parser();
		"(testing;)", // invalid semicolon
		"(1 +)", // missing rhs
		"( / 888)", // missing rhs
		// "((testing)", // unclosed inner group
	);
}

#[test]
fn recover_block() {
	insta_assert_parser!(
		expr_parser();
		"{ testing }", // missing semicolon
		// TODO fix unclosed delimiters
		// "{ (testing; }", // unclosed group
		// "{ {testing; }", // unclosed block
	);
}

#[test]
fn recover_call() {
	insta_assert_parser!(
		expr_parser();
		"testing(,)", // missing args
		"testing(,)()()", // missing args
		"testing(1 2 3)", // missing commas
		"testing(1 + / 2)", // missing commas
		// "testing(1, {2, 3)", // unclosed block
	);
}
