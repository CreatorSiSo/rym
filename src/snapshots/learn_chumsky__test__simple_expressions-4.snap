---
source: src/test.rs
expression: "&snapshot"
---
--- Input ---
(make_func())()()()
---

Some(
    (
        Call(
            (
                Call(
                    (
                        Call(
                            (
                                Call(
                                    (
                                        Ident("make_func"),
                                        1..10,
                                    ),
                                    [],
                                ),
                                1..12,
                            ),
                            [],
                        ),
                        1..15,
                    ),
                    [],
                ),
                1..17,
            ),
            [],
        ),
        1..19,
    ),
)

--- Errors ---
[]
---
